File: Stacks/stack.java
// Implemenmtation of  stack data structure
/*
    This implementation uses a generic type T to allow the stack to store elements of any type. 
    The stack is implemented using an array and has methods for pushing, popping, peeking, checking 
    if the stack is empty, and getting the size of the stack. The comments explain each step of the 
    implementation.
*/
public class Stack<T> {
    // array to store elements of the stack
    private T[] elements;
    // top index of the stack
    private int top;

    // constructor to initialize the stack with a given capacity
    public Stack(int capacity) {
        // create a new array of type T with the given capacity
        elements = (T[]) new Object[capacity];
        // initialize top index to -1
        top = -1;
    }

    // push an element onto the top of the stack
    public void push(T element) {
        // check if the stack is full
        if (top == elements.length - 1) {
            // if the stack is full, throw an exception
            throw new RuntimeException("Stack overflow");
        }
        // increment top index
        top++;
        // insert the element at the top index
        elements[top] = element;
    }

    // pop the top element from the stack and return it
    public T pop() {
        // check if the stack is empty
        if (top == -1) {
            // if the stack is empty, throw an exception
            throw new RuntimeException("Stack underflow");
        }
        // get the top element
        T element = elements[top];
        // decrement top index
        top--;
        // return the top element
        return element;
    }

    // return the top element of the stack without removing it
    public T peek() {
        // check if the stack is empty
        if (top == -1) {
            // if the stack is empty, throw an exception
            throw new RuntimeException("Stack underflow");
        }
        // return the top element
        return elements[top];
    }

    // return true if the stack is empty, false otherwise
    public boolean isEmpty() {
        return top == -1;
    }

    // return the size of the stack
    public int size() {
        return top + 1;
    }
}

File: Stacks/stack.cpp
// Implemenmtation of  stack data structure
/*
    This implementation creates a class Stack that has a private integer top to keep track of the index 
    of the topmost element in the stack, and an integer array arr to store the elements of the stack. 
    The push, pop, peek, and isEmpty methods are used to add elements to, remove elements from, view 
    the topmost element in, and check if the stack is empty, respectively. The main function creates 
    an instance of Stack and demonstrates the use of the stack by pushing some elements onto it, 
    printing the topmost element, popping an element from the stack, and checking if the stack is empty.
*/
#include <iostream>
using namespace std;

const int MAX_SIZE = 100; // Maximum size of stack

class Stack {
private:
    int top; // Index of topmost element in stack
    int arr[MAX_SIZE]; // Array to store elements of stack
public:
    Stack() {
        top = -1; // Initialize top index to -1, indicating empty stack
    }

    // Pushes an element onto the top of the stack
    void push(int x) {
        if (top == MAX_SIZE - 1) { // Check if stack is full
            cout << "Stack Overflow\n";
            return;
        }
        arr[++top] = x; // Increment top index and add element to array
    }

    // Removes and returns the topmost element from the stack
    int pop() {
        if (top == -1) { // Check if stack is empty
            cout << "Stack Underflow\n";
            return -1;
        }
        int val = arr[top]; // Store topmost element in a variable
        top--; // Decrement top index to remove element from stack
        return val; // Return topmost element
    }

    // Returns the topmost element in the stack without removing it
    int peek() {
        if (top == -1) { // Check if stack is empty
            cout << "Stack Underflow\n";
            return -1;
        }
        return arr[top]; // Return topmost element
    }

    // Returns true if stack is empty, false otherwise
    bool isEmpty() {
        return top == -1;
    }
};

int main() {
    Stack s;

    // Push some elements onto the stack
    s.push(5);
    s.push(10);
    s.push(15);

    // Print the topmost element without removing it
    cout << "Top element: " << s.peek() << endl;

    // Pop an element from the stack and print it
    cout << "Popped element: " << s.pop() << endl;

    // Check if stack is empty
    cout << "Is stack empty? " << (s.isEmpty() ? "Yes" : "No") << endl;

    return 0;
}
